### PR TITLE
Initial API implementation

### PR DESCRIPTION
**Pull request by plusmouse**
Done

### CODE REVIEW

File: Source/Utilities/VerifyListTypes.lua [hunk 1]
```diff
@@ -0,0 +1,18 @@
+-- array: List of objects
+-- typeStr: Type to ensure every item is
+--
+-- Returns a new array with only valid entries
+function Auctionator.Utilities.VerifyListTypes(array, typeStr)
```
Comment (jrob8577) [inline Source/Utilities/VerifyListTypes.lua]:
```lua
function Auctionator.Utilities.VerifyListTypes(list, requiredTypeString)
  if list == nil or type(list) ~= "table" then
    return nil
  end

  local result
  for _, item in ipairs(list) do
    if type(item) == requiredTypeString then
      table.insert(result, item)
    else
      return nil
    end
  end

  return result
end
```

File: Source/Utilities/ItemKeyFromLink.lua [hunk 2]
```diff
@@ -1,15 +1,14 @@
...
   if itemLink ~= nil then
     local _, _, itemString = string.find(itemLink, "^|c%x+|H(.+)|h%[.*%]")
-    local linkType, itemId, _, _, _, _, _, _, _ = strsplit(":", itemString)
-    if linkType == "battlepet" then
-      return "p:"..itemId;
-    elseif linkType == "item" then
-      return itemId;
-    else
-      return nil
+    if itemString then
```
Comment (jrob8577) [inline Source/Utilities/ItemKeyFromLink.lua]: Use explicit comparison value (i.e. `itemString ~= nil`)

File: Source/API/README.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+Calling any other functions related to Auctionator is not supported and may
+break without warning in future releases.
+
+This API is intended to remain stable.
+
+```lua
+-- Returns the last scanned price for an item identified by itemID
+-- Returns the price in coppers, or nil if one wasn't found
+Auctionator.API.v1.GetAuctionPriceByItemID(callerID, itemID)
```
Comment (jrob8577) [inline Source/API/README.md]: We should document the parameters as well, and provide a sample call with sample results, for clarity.

File: Source/API/MultiSearch.lua [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  --Validate arguments
+  local cloned = Auctionator.Utilities.VerifyListTypes(searchTerms, "string")
+  if not cloned then
+    Auctionator.API.ComposeError(
+      callerID, "Usage Auctionator.API.v1.MultiSearch(string[])"
+    )
+  end
+  -- Validate state
+  if not AuctionHouseFrame or not AuctionHouseFrame:IsShown() then
```
Comment (jrob8577) [inline Source/API/MultiSearch.lua]: Should we do this check before bothering to assemble the list?

File: Source/API/Internal.lua [hunk 5]
```diff
@@ -0,0 +1,13 @@
+function Auctionator.API.ComposeError(callerID, message)
+  error(
+    "Contact the maintainer of " .. callerID ..
+    " to resolve this problem. Details: " .. message
+  )
+end
+
+-- TODO: Maintain authorization keys for add-ons to prevent false callerIDs
```
Comment (jrob8577) [inline Source/API/Internal.lua]: Note for ourselves, we should probably store encrypted values and encrypt the provided string so people can't just use other addon's names, if we go this route

Comment (plusmouse) [reply Source/API/Internal.lua]: Yeah, for now its just a "fear factor", but it is something for the future

File: Source/API/MultiSearch.lua (hunk 4 above)
Comment (plusmouse) [reply Source/API/MultiSearch.lua]: I just went with verify parameters, then verify state

Comment (plusmouse) [reply Source/API/MultiSearch.lua]: But I don't think there's any benefit to doing it earlier, either way they have to fix both problems

Comment (jrob8577) [reply Source/API/MultiSearch.lua]: Makes sense.

**Pull request merged**